src/main/java/com/google/sps/servlets/PotentialMatchesServlet.java

```diff
@@ -0,0 +1,110 @@
+*/
+@WebServlet("/potential-matches")
+public class PotentialMatchesServlet extends HttpServlet {
+  private static final String USER_ENTITY = "User";
+  private static final String USER_ID_PROPERTY = "id";
+  private static final String USER_FRIENDS_LIST_PROPERTY = "friends-list";
+  private static final String NO_POTENTIAL_MATCH_RESULT = "NO_POTENTIAL_MATCHES";
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String currUserID = request.getParameter("userid");
+    
+    String nextPotentialMatchID = loadUserPotentialMatch(currUserID);
+
+    response.setContentType("text/html");
+    response.getWriter().print(nextPotentialMatchID);
+  }
+
+  /**
+  * Loads the next potential match for a user
+  *
+  * <p>Currently is just loading all the possible matches and obtaining one of them.
+  *
+  * <p>TODO(#21): Obtain potential match results from datastore instead of running the 
+  * potential match finding methods repeatedly.
+  *
+  * @param userID The userID of the user who's potential match is being found
+  * @return The next potential match for a user
+  */
+  private String loadUserPotentialMatch(String userID) {
+    ImmutableSet<UserNode> userNodes = createUserNodes();
+    UserFriendsMap friendsMap = new UserFriendsMap(userNodes);
+    
+    ImmutableSet<String> potentialMatches = PotentialMatchAlgorithm.findPotentialMatchesForUser(userID, friendsMap);
+  
+    String nextPotentialMatchID = potentialMatches.size() > 0 ?
+      potentialMatches.iterator().next() : NO_POTENTIAL_MATCH_RESULT;
+    
+    return nextPotentialMatchID;
+  }
+
+  /**
+  * Helper method that creates the set of UserNodes that will be fed into the potential
+  * matching method
+  *
+  * @return the set of user nodes for all current app users
+  */
+  private ImmutableSet<UserNode> createUserNodes() {
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(new Query(USER_ENTITY));
+    List<Entity> entityResults = results.asList(FetchOptions.Builder.withDefaults());
+
+    ImmutableSet.Builder<UserNode> builder = ImmutableSet.builder();
+    for (Entity userEntity: entityResults) {
+      String userID = (String) userEntity.getProperty(USER_ID_PROPERTY);
+      ArrayList<String> friendsIds =
+        (ArrayList<String>) userEntity.getProperty(USER_FRIENDS_LIST_PROPERTY);
+      UserNode userNode = new UserNode(userID,
+        friendsIds != null ? ImmutableSet.copyOf(friendsIds) : ImmutableSet.of());
+      builder.add(userNode);
+    }
+
+    ImmutableSet<UserNode> allUserNodes = builder.build();
+    
+    return allUserNodes;
+  }
+
+/** Wrap the doGet method for jUnit testing */
+  public void doGetWrapper(DatastoreService datastoreService, HttpServletRequest request, HttpServletResponse response) throws IOException {
+    DatastoreService datastore = datastoreService;
```

Comment:
How is this working? You're manually creating your `DatastoreService` inside of `createUserNode()`, so it seems like this wouldn't be used.

Reply:
Oh you're right! I was following along with Jess's servlet testing to help write the servlet tests for this and used a doGetWrapper method, but realizing now with the way I wrote the methods in my servlet that I don't need to keep the wrapper.